In [1]:
import math
math.sqrt(16)

4.0

# Approach: Module import and invocation

## Problem Explanation

Calculates square root of a number using Python's standard `math` module.

## Approach Used

* **Technique:** Module import and invocation.
* **Why it works:** Accesses math library function `sqrt()`.
* **Key Idea:** Math standard library functions.

## Dry Run

Input: `math.sqrt(16)`
* Evaluates `√16 = 4.0`.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Math functions execute in constant time.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; No extra storage.
</div>

In [4]:
from math import sqrt,pi
print(sqrt(16))
print(sqrt(25))
print(pi)


4.0
5.0
3.141592653589793


# Approach: Specific symbol import

## Problem Explanation

Demonstrates importing specific definitions (`sqrt`, `pi`) from the `math` module directly into the global namespace.

## Approach Used

* **Technique:** Specific symbol import.
* **Why it works:** Directly pulls requested attributes.
* **Key Idea:** Cleaner namespaces.

## Dry Run

Outputs:
* `sqrt(16)` -> `4.0`
* `sqrt(25)` -> `5.0`
* `pi` -> `3.14159...`

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Direct execution.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Minimal namespace symbols.
</div>

In [5]:
import numpy as np
np.array([1,2,3,4,5])


array([1, 2, 3, 4, 5])

# Approach: Library initialization

## Problem Explanation

Imports `numpy` and constructs a basic numerical array structure.

## Approach Used

* **Technique:** Library initialization.
* **Why it works:** Wraps standard python list in numpy array.
* **Key Idea:** Vector operations foundation.

## Dry Run

Input: `[1, 2, 3]`
* Yields: `numpy.array([1, 2, 3])`.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(N)</span> &mdash; Copies list elements into arrays.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(N)</span> &mdash; Storage for numpy array.
</div>

In [17]:
## random
import random
print(random.randint(1,10))
print(random.choice(['apple','banana','cherry']))

3
cherry


# Approach: Randomization algorithms

## Problem Explanation

Generates pseudo-random choices and random integers using Python's `random` module.

## Approach Used

* **Technique:** Randomization algorithms.
* **Why it works:** Accesses pseudo-random number generator routines.
* **Key Idea:** Random value sampling.

## Dry Run

Outputs:
* `randint(1, 10)` -> random integer between 1 and 10 (inclusive).
* `choice(...)` -> random fruit element from the list.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Standard random generation uses constant time.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Minimal state variables.
</div>

In [19]:
#file and directory access
import os
print(os.getcwd())

os.mkdir('test_dir')

c:\users\vaidik\Downloads\dsa\DSA_python


# Approach: OS operations file handling

## Problem Explanation

Interacts with the operating system to print the current working directory and create a new directory.

## Approach Used

* **Technique:** OS operations file handling.
* **Why it works:** Invokes OS system calls through standard library API.
* **Key Idea:** System path management.

## Dry Run

* `os.getcwd()` returns current directory string.
* `os.mkdir('test_dir')` makes directory under working folder.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; System filesystem operations.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Path directory references.
</div>

In [21]:
##high kevel operation on files and collection of files
import shutil
shutil.copyfile('source.txt','destination.txt')

'destination.txt'

# Approach: High-level file management operations

## Problem Explanation

Copies a file named `source.txt` to `destination.txt` using the `shutil` module.

## Approach Used

* **Technique:** High-level file management operations.
* **Why it works:** Copies bytes from source file path to destination path.
* **Key Idea:** File duplication.

## Dry Run

* Checks `source.txt` and copies contents to `destination.txt`.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(B)</span> &mdash; Proportional to size in bytes of source file.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Internal buffer space is constant.
</div>

In [27]:
#data serialization
import json
data={'name':'vaidik','age':22}

json_str=json.dumps(data)
print(json_str)
print(type(json_str))

parsed_data=json.loads(json_str)
print(parsed_data)
print(type(parsed_data))

{"name": "vaidik", "age": 22}
<class 'str'>
{'name': 'vaidik', 'age': 22}
<class 'dict'>


# Approach: Data serialization

## Problem Explanation

Demonstrates JSON string serialization and deserialization using the standard `json` module.

## Approach Used

* **Technique:** Data serialization.
* **Why it works:** Uses `json.dumps()` to parse Python dictionary to JSON string, and `json.loads()` to deserialize string back to dict.
* **Key Idea:** Python object data interchange representation.

## Dry Run

Input: `data={'name':'vaidik','age':22}`
* `dumps` makes string `'{"name": "vaidik", "age": 22}'`.
* `loads` returns dictionary object.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(C)</span> &mdash; Linear with respect to character length of representation.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(C)</span> &mdash; Storage for JSON strings and structures.
</div>

In [33]:
##csv
import csv
with open('example.csv',mode='w',newline='') as file:
    writer = csv.writer(file)
    writer.writerow(['name','age'])
    writer.writerow(['vaidik','25'])

with open('example.csv',mode='r') as file:
    reader =  csv.reader(file)
    for row in reader:
        print(row)


['name', 'age']
['vaidik', '25']


# Approach: CSV operations

## Problem Explanation

Demonstrates writing data records to a CSV file and reading them back using Python's `csv` module.

## Approach Used

* **Technique:** CSV operations.
* **Why it works:** Employs `csv.writer` to write structured rows, and `csv.reader` to iterate lines.
* **Key Idea:** Flat database parsing.

## Dry Run

* Writes header `['name','age']` and data `['vaidik','25']`.
* Reads file and prints list representation of rows.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(R * C)</span> &mdash; R rows of C columns.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(R * C)</span> &mdash; To store row data read in memory.
</div>

In [ ]:
##datetime
from datetime import datetime, timedelta
now=datetime.now()
print(now)

yesterday = now-timedelta(days=1)
print(yesterday)

2026-06-25 15:51:56.876749
2026-06-23 15:51:56.876749


# Approach: Date arithmetic operations

## Problem Explanation

Calculates current timestamp and yesterday's date using `datetime` and `timedelta` objects.

## Approach Used

* **Technique:** Date arithmetic operations.
* **Why it works:** Subtracts `timedelta(days=1)` interval from current time reference.
* **Key Idea:** Calendar date offsets.

## Dry Run

* Gets current system time.
* Subtracts 1 day interval to output yesterday's date.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Date adjustments run in constant time.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Calendar record representations.
</div>

In [125]:
#time
import time 
print(time.time())
time.sleep(2)
print(time.time())
time.sleep(2)
print(time.time())
time.sleep(2)
print(time.time())
time.sleep(2)


1782383252.332583
1782383254.333766
1782383256.3346524
1782383258.3355994


# Approach: Execution delays

## Problem Explanation

Uses `time.sleep()` to block program execution for designated second intervals.

## Approach Used

* **Technique:** Execution delays.
* **Why it works:** Halts process threads for the specified time durations.
* **Key Idea:** Timer sleep interrupts.

## Dry Run

* Prints timestamp, waits 2 seconds, prints next timestamp, repeating.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(D)</span> &mdash; Blocks execution proportional to sleep delay D.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Thread sleep timers.
</div>

In [129]:
#regular expression
import re
pattern = r'\d+' #use to find digits
text = 'there are 123 apples'
match=re.search(pattern,text)
print(match.group())

123


# Approach: Regex search match

## Problem Explanation

Extracts the first sequence of digits from a string using regular expressions.

## Approach Used

* **Technique:** Regex search match.
* **Why it works:** Evaluates the regex pattern `\d+` which matches one or more consecutive digit characters.
* **Key Idea:** String search patterns.

## Dry Run

Input: `'there are 123 apples'`
* Scan characters, finds matching digits: `'123'`.
* Output: `'123'`.

## Time Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Time Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(L)</span> &mdash; Proportional to input string length L.
</div>

## Space Complexity

<div style="padding: 8px 12px; border-left: 4px solid #5cb85c; background-color: #dff0d8; color: #3c763d; margin: 10px 0;">
  <strong>Space Complexity:</strong> <span style="color: #5cb85c; font-weight: bold;">O(1)</span> &mdash; Regex engine memory.
</div>